### PR TITLE
[Feature]:add dynamic shape support for two_stage func

### PR DESCRIPTION
**Pull request by tangyanf**
this pr was useful for exported onnx model support dynamic input shapes, there are four changes in total:
1. rpn_head.py, modify topk's attribute K, which was used to keep topk op for dynamic K in onnx model, use                           
    
     ```python
         nms_pre = torch.where(scores_shape[0] < cfg.nms_pre, scores_shape[0], cfg.nms_pre)
     ```

     to replace                                                          
 
    ```python
       if scores.shape[0] > cfg.nms_pre: 
           nms_pre = scores.shape[0] 
       else:
           nms_pre = cfg.nms_pre
    ```
                                                                                                  
2. two_stage.py, add code
    ```python
        if torch.onnx.is_in_onnx_export():
            img_shape = torch._shape_as_tensor(img)[2:]
            img_metas[0]['img_shape_for_onnx'] = img_shape
    ```
    keep the original input shape for Post-processing which was related to calculation of dynamic shapes
3. single_stage.py, same as above
4. test_mixins.py, add code
    ````python
        if torch.onnx.is_in_onnx_export():
            img_shapes = tuple(meta['img_shape_for_onnx']
                               for meta in img_metas)
        else:
            img_shapes = tuple(meta['img_shape'] for meta in img_metas)
    ````
    get original input shape to keep dynamic in exporting onnx model and use shape() to replace len()

    ```python
        num_proposals_per_img = tuple(p.shape[0] for p in proposals)
    ```
    to replace

    ```python
        num_proposals_per_img = tuple(len(p) for p in proposals)
    ```

this pr was mainly for ONNX export and only tested on CPU.

### CODE REVIEW

File: mmdet/models/detectors/two_stage.py [hunk 1]
```diff
@@ -190,6 +190,10 @@ def simple_test(self, img, img_metas, proposals=None, rescale=False):
 
         x = self.extract_feat(img)
 
+        # get origin input shape to onnx dynamic input shape
+        if torch.onnx.is_in_onnx_export():
+            img_metas[0]['img_shape_for_onnx'] = img.shape[2:]
```
Comment (RunningLeon) [inline mmdet/models/detectors/two_stage.py:195]: Argument `img_shape` for function `delta2bbox` expects type as tensor. Try as following:
```
        # get origin input shape to support onnx dynamic shape
        if torch.onnx.is_in_onnx_export():
            # get shape as tensor
            img_shape = torch._shape_as_tensor(img)[2:]
            img_metas[0]['img_shape_for_onnx'] = img_shape
```

**Comment by RunningLeon** [comment]
@tangyanf  We could add the short description of this PR at the beginning.

> description

**Comment by tangyanf** [comment]
> @tangyanf We could add the short description of this PR at the beginning.
> 
> > description

ok，i will add it

File: mmdet/models/detectors/two_stage.py (hunk 1 above)
Comment (tangyanf) [reply mmdet/models/detectors/two_stage.py:195]: > Argument `img_shape` for function `delta2bbox` expects type as tensor. Try as following:
> 
> ```
>         # get origin input shape to support onnx dynamic shape
>         if torch.onnx.is_in_onnx_export():
>             # get shape as tensor
>             img_shape = torch._shape_as_tensor(img)[2:]
>             img_metas[0]['img_shape_for_onnx'] = img_shape
> ```

that's my fault, i'll chek it

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by RunningLeon** [comment]
@ZwwWayne Check this PR and hopefully merge it soon.

**Comment by ZwwWayne** [comment]
1. Please describe the motivation, modification, and outcome of this PR in detail. For example, which functions are modified for what, what is the influence.
2. Please indicate whether this is tested in GPU mode and whether related models are tested on GPU without abnormal accuracy, as this is mainly for ONNX export and may only tested on CPU.

File: mmdet/models/dense_heads/rpn_head.py [hunk 2]
```diff
@@ -134,20 +136,26 @@ def _get_bboxes_single(self,
                 scores = rpn_cls_score.softmax(dim=1)[:, 0]
             rpn_bbox_pred = rpn_bbox_pred.permute(1, 2, 0).reshape(-1, 4)
             anchors = mlvl_anchors[idx]
-            if cfg.nms_pre > 0 and scores.shape[0] > cfg.nms_pre:
+            if cfg.nms_pre > 0:
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:139]: Will this change affect model's GPU inference?

**Review by ZwwWayne — CHANGES_REQUESTED** [review]

File: mmdet/models/dense_heads/rpn_head.py (hunk 2 above)
Comment (tangyanf) [reply mmdet/models/dense_heads/rpn_head.py:139]: model's gpu inference will go through the elif branch, and it will not affect the original logic

Comment (tangyanf) [reply mmdet/models/dense_heads/rpn_head.py:139]: > Will this change affect model's GPU inference?

model's gpu inference will go through the elif branch, and it will not affect the original logic

**Review by ZwwWayne — APPROVED** [review]
